### ISSUE TITLE
Find by primary key using named/typed parameters

### ISSUE THREAD

**Issue by dvoreckyaa** [dotnet/efcore#30587]
Hi all,
we have possibility to find an entity usig Find methdod

```cs
public virtual TEntity? Find(params object?[]? keyValues)
```

where we have to pass all paramters as object array. For me it seems a bit inconvenient and can lead to mistakes.
**I would suggest improving this by adding FindByPrimaryKey methods that will accept named and typed parameters.**
Maybe this functionality is already implemented but I could not find.

```cs 
public static Db.Models.ItsDetail FindByPrimaryKey(this DbSet<Models.ItsDetail> entities, Int32 detWstagNo, Int32 detNo)
{
    return entities.Find(detWstagNo, detNo);
}
```

Actually, a few years ago I implemented a generator that creates an extension for a DbSet to support FindByPK with typed parameters.

https://github.com/dvoreckyaa/FindByPKGenerator
![](https://github.com/dvoreckyaa/FindByPKGenerator/blob/main/img/screenshot1.png?raw=true)

**Comment by ajcvickers** [comment]
@dvoreckyaa Do you have a general purpose API proposal here? The examples above all seem specific to certain entity type with known key configurations.

**Comment by dvoreckyaa** [comment]
@ajcvickers , not it is not specific to certain entity type. The mentioned generator takes a DbContext (should be inherited fom DbContext) then

1. Calls GetEntityTypes (current implementations creates DbContext.UseInMemoryDatabase)
2. Lists all selected object types and calls FindPrimaryKey() for each type.
3. Based on the found PK, creates FindByPK and FindByPKAsync using tt.

```cs
public void GenerateFileFromType<T>(string outputFolder) where T : DbContext
```

**Comment by roji** [comment]
@dvoreckyaa if I understand your proposal correctly, this would be a source generator provided by EF Core, which generates Find extension methods over concrete `DbSet<T>`, accepting the specific primary key property (or property types). For example, if I have a Blog entity type with a Guid key, the source generator would generate an extension method over `DbSet<Blog>` which accepts a single Guid.

A source generator like this would need to access the user's EF model, in order to know about entities and their primary key properties. There are unfortunately various problems that make it unfeasible for a source generator to know about the model, see https://github.com/dotnet/efcore/issues/25009#issuecomment-853735001 for more details.

In any case, we've rarely seen any actual user complaints about bugs because of this, and it's fairly trivial for users to add their own extensions manually, if they so wish. There's also the option to just use a LINQ query to load an entity type from the database (i.e. `ctx.Blogs.Single(b => b.Id == id)`), which is fully strongly-typed and safe. This doesn't first check the change tracker like Find, though that can also be done by accessing [ChangeTracker.Entries](https://learn.microsoft.com/en-us/ef/core/change-tracking/entity-entries#using-changetrackerentries-to-access-all-tracked-entities).

To summarize, in an ideal world something like this would indeed exist; but the significant issues in actually doing this, coupled with the relative low value and alternative methods for achieving almost the same, make this pretty low value in my opinion.

**Comment by dvoreckyaa** [comment]
@roji, it sounds reasonable, thanks for the clarification.
